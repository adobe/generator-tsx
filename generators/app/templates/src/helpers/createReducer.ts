import { Reducer } from 'redux'
import { DeepReadonly } from 'ts-essentials'

import RootActions from 'actions'

export default function createReducer<T>(name: string, defaultState: T) {
	return <A extends RootActions>(
		callback: (state: DeepReadonly<T>, action: A) => DeepReadonly<T>,
	) => {
		const reducer: Reducer<DeepReadonly<T>, A> = (
			previousState = defaultState as DeepReadonly<T>,
			action: A,
		) => callback(previousState, action)
		return Object.defineProperty(reducer, 'name', {
			value: `${name}Reducer`,
			configurable: true,
		}) as typeof reducer
	}
}
