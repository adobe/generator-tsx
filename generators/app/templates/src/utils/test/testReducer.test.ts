import { AnyAction } from 'redux'
import { DeepReadonly } from 'ts-essentials'

import createReducer from 'helpers/createReducer'

import testReducer from './testReducer'

const reducer = createReducer('test', { foo: 'INITIAL_STATE' })(
	(state, action: AnyAction) =>
		action.type === 'UPDATE_FOO' ? { ...state, foo: action.value } : state,
)

testReducer(reducer, test => {
	const newValue = 'NEW_VALUE'
	const expectedState = { foo: newValue }
	const action = { type: 'UPDATE_FOO', value: newValue }

	test(null, action, expectedState)
	test(null, action, expectedState, 'handles expectedState')
	test(null, action, onNewState, 'handles an expected state callback')

	function onNewState(newState: DeepReadonly<{ foo: string }>) {
		expect(newState).toEqual(expectedState)
	}
})
