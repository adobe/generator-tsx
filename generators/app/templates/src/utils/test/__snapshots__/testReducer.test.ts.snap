// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testReducer returns initial state 1`] = `
Object {
  "foo": "INITIAL_STATE",
}
`;
